### PR TITLE
Added primary key to Person as our first internal state field to track and save

### PR DESCRIPTION
**Pull request by podocarp**
- Created new InternalState* classes to track and store the internal state of the program
- Examples of internal states:
  - Command history
  - Primary keys of other thingawumuts
  - etc.

The UserPrefs now store one more path, that is the path of the internal state save file. By default it is `data/state.json`.

The only things that need to be touched when adding more fields in is the `{update, apply}InternalState` functions. Be sure to update them!

### CODE REVIEW

File: src/main/java/seedu/address/model/InternalState.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package seedu.address.model;
+
+import seedu.address.model.person.Person;
+
+/**
+* This class holds the actual state of the application.*/
+public final class InternalState {
+    private int primaryKeyCounter;
```
Comment (daekoon) [inline src/main/java/seedu/address/model/InternalState.java]: Can we change the variable name to be more specific to `Person` class? We eventually have to add a `primaryKeyCounter` for `Activity` class as well.

File: src/main/java/seedu/address/storage/AddressBookStorage.java [hunk 2]
```diff
@@ -8,38 +8,36 @@
...
+    * Saves the given {@link ReadOnlyAddressBook} to the storage.
+    * @param addressBook cannot be null.
+    * @throws IOException if there was any problem writing to the file.
+    */
     void saveAddressBook(ReadOnlyAddressBook addressBook) throws IOException;
 
     /**
-     * @see #saveAddressBook(ReadOnlyAddressBook)
-     */
+    * @see #saveAddressBook(ReadOnlyAddressBook)
```
Comment (daekoon) [inline src/main/java/seedu/address/storage/AddressBookStorage.java:40]: I'm surprised these passes checkStyle o.o

**Review by daekoon — CHANGES_REQUESTED** [review]
Mostly ok, but we should at least do this before merging:
- Update test helper classes `PersonBuilder`. Add in a method to set `primaryKey`.
- Change the naming of the `primaryKeyCounter` to be more specific towards `Person` class. 

These should be future sprint enhancements:
- Tests for `InternalStorage` and related classes (low priority)
- Validation to check if `primaryKeyCounter` for `InternalStorage` is valid upon file load. (Higher priority)

File: src/main/java/seedu/address/storage/AddressBookStorage.java (hunk 2 above)
Comment (podocarp) [reply src/main/java/seedu/address/storage/AddressBookStorage.java:40]: please ignore, they are just fixes by prettier and yea checkstyle doesn't see them

**Comment by podocarp** [comment]
>     * Update test helper classes `PersonBuilder`. Add in a method to set `primaryKey`.
primarykey being exposed is quite contentious. Some points about the behavior now:

- `Person.equals` does not care about primary key.
- `PersonBuilder` constructor can be made to copy the primary key.
- `PersonBuilder.build()` calls a constructor of `Person` which does not care about the primary key. 

So it is quite messy. I will add some tests for other behavior, but I don't think I will ever want to change the `equals` behavior. For sure the primary key will be different, what's there to check (it's validated on start)? And so it follows that if `equals` behaves like this, the other functions can do anything they want, but nothing will change. e.g. `PersonBuilder.build()` can use any constructor it wants and nothing will change.

>     * Tests for `InternalStorage` and related classes (low priority)
Done

>     * Validation to check if `primaryKeyCounter` for `InternalStorage` is valid upon file load. (Higher priority)
Done

**Review by daekoon — APPROVED** [review]

**Review by daekoon — APPROVED** [review]
Good job for catching those edge cases regarding `.equal` method for `Person` class. Should document it in the Developer guide in the future.
Personally I think your `equal` method is ok - it should be comparing the content of the `Person` object, and I would argue that `primaryKey` isn't really a content , but an identity of that object and hence shouldn't be used for `.equal` comparison (similar to how hashcode doesn't get compared)

**Review by liakify — APPROVED** [review]
Looks good to me

**Pull request merged**